### PR TITLE
windows: relax "safe" permissions for binary execution

### PR DESCRIPTION
**Pull request by muffins**
The "safe" permissions for the Windows binaries were a little too restrictive to facilitate a good user experience in deploying and running the osquery service. This PR relaxes the permissions as detailed below and extends our Powershell scripting with helper functions to assure "safe" permissions regardless of a priori conditions on the directory.

As a note - **documentation** for this behavior as well as detailed osquery service installation instructions is (will be updated) contained in  #3705.

In order for an osquery binary, daemon or extension, to be considered "safe", the following must hold true:

1. The owner of the binary and containing parent directory is the `Administrators` group
2. No lower privileged user account has `Write` permissions on either the binary or the containing parent directory.
3. (Optional) the `SYSTEM` account may have full permissions on the binary and the containing parent directory.

The helper function we bundle in the `osquery_utils.ps1` script does the following to ensure "safe" permissions exist on the directory specified:

1. Removes all pre-existing ACL information from the file, as well as sub-directories and files, and disables inheritance
2. Grants the `Administrators` and `SYSTEM` accounts Full access to the directory and sub-folders using container and object inheritance.
3. Grants the `Users` built-in group Read and Execute permissions on the directory and sub-folders using container and object inheritance.
4. Sets the owner of the directory, sub-folders, and files to the `Administrators` group.

The above steps ensure that "safe" permissions are in place.  The old behavior of leveraging an explicit `Deny-Write` ACE for the `Everyone` group is still supported, however we now only require ownership by `Administrators` and no low privileged write access.

### CODE REVIEW

**Comment by theopolis** [comment]
This looks intense, is there another Windows developer around that can take a look?

**Comment by theopolis** [comment]
Remember you don't want `Users` to have `Read` on the `osquery.db` folder.

File: osquery/filesystem/windows/fileops.cpp [hunk 1]
```diff
@@ -375,18 +375,16 @@ static Status hasAccess(HANDLE handle, mode_t mode) {
...
 
-  status =
-      ::GetUserObjectSecurity(handle, &security_info, nullptr, 0, &sd_size);
-  if (status || (!status && ::GetLastError() != ERROR_INSUFFICIENT_BUFFER)) {
+  status = GetUserObjectSecurity(handle, &security_info, nullptr, 0, &sd_size);
+  if (status || (!status && GetLastError() != ERROR_INSUFFICIENT_BUFFER)) {
     return Status(-1, "GetUserObjectSecurity get SD size error");
   }
 
   std::vector<char> sd_buffer;
```
Comment (yying) [inline osquery/filesystem/windows/fileops.cpp]: Consolidate this info `std::vector<char> sd_buffer(sd_size, '\0');` similar to the others

**Comment by muffins** [comment]
The most recent commit doesn't include the note from @theopolis regarding User DB permissions and I haven't tested if these are implicitly gained - but I will soon!

**Comment by osqueryer** [comment]
:no_good: The commit aefbfe8377613a7a342042a8dd93611129b6dc8a (Job results: [2555](https://jenkins.osquery.io/job/osqueryPullRequestAudit/2555/)) failed the code audit or documentation test.

**Comment by osqueryer** [comment]
:-1: The commit aefbfe8377613a7a342042a8dd93611129b6dc8a (Job results: [2778](https://jenkins.osquery.io/job/osqueryPullRequestWindows/2778/)) failed one or more tests (**Windows**).

**Comment by osqueryer** [comment]
:-1: The commit aefbfe8377613a7a342042a8dd93611129b6dc8a (Job results: [818](https://jenkins.osquery.io/job/osqueryPullRequestFreeBSD/818/)) failed one or more tests (**FreeBSD**).

**Comment by muffins** [comment]
Tagging `in-progress` until I can assert that this checks the DB permissions as well.

**Comment by osqueryer** [comment]
:no_good: The commit 5e9963c0db0364d35118ea518fcf008169e74eac (Job results: [2627](https://jenkins.osquery.io/job/osqueryPullRequestAudit/2627/)) failed the code audit or documentation test.

**Comment by osqueryer** [comment]
:-1: The commit 5e9963c0db0364d35118ea518fcf008169e74eac (Job results: [2861](https://jenkins.osquery.io/job/osqueryPullRequestWindows/2861/)) failed one or more tests (**Windows**).

**Comment by osqueryer** [comment]
:-1: The commit 2f088da4e7fd1756ca2713f4e3b329a045b6e199 (Job results: [2864](https://jenkins.osquery.io/job/osqueryPullRequestWindows/2864/)) failed one or more tests (**Windows**).

**Comment by osqueryer** [comment]
:-1: The commit 2f088da4e7fd1756ca2713f4e3b329a045b6e199 (Job results: [2866](https://jenkins.osquery.io/job/osqueryPullRequestWindows/2866/)) failed one or more tests (**Windows**).

File: osquery/filesystem/windows/fileops.cpp [hunk 2]
```diff
@@ -501,47 +496,49 @@ static AclObject modifyAcl(PACL acl,
...
+    if (!GetAce(acl, i, &void_ent)) {
       return std::move(AclObject());
     }
 
+    auto entry = static_cast<PACE_HEADER>(void_ent);
     if ((entry->AceFlags & INHERITED_ACE) == INHERITED_ACE) {
       break;
     }
 
+    auto allowed_ace = reinterpret_cast<ACCESS_ALLOWED_ACE*>(entry);
```
Comment (muffins) [inline osquery/filesystem/windows/fileops.cpp]: Gunna move these down closer to where they're actually used to avoid the `cast`ing if possible.

**Comment by osqueryer** [comment]
:no_good: The commit 6f3d500b1fc2f2e838e9bfe08a37f130be3b8453 (Job results: [2640](https://jenkins.osquery.io/job/osqueryPullRequestAudit/2640/)) failed the code audit or documentation test.

**Comment by osqueryer** [comment]
:-1: The commit 779ca7799a71ebe84822b7e3dc02d922f14b2c54 (Job results: [911](https://jenkins.osquery.io/job/osqueryPullRequestFreeBSD/911/)) failed one or more tests (**FreeBSD**).

**Comment by osqueryer** [comment]
Build finished. 
:-1: The commit 779ca7799a71ebe84822b7e3dc02d922f14b2c54 (Job results: [912](https://jenkins.osquery.io/job/osqueryPullRequestFreeBSD/912/)) failed one or more tests (**FreeBSD**).

**Comment by osqueryer** [comment]
:-1: The commit 779ca7799a71ebe84822b7e3dc02d922f14b2c54 (Job results: [912](https://jenkins.osquery.io/job/osqueryPullRequestFreeBSD/912/)) failed one or more tests (**FreeBSD**).

**Comment by muffins** [comment]
ok to test

**Comment by osqueryer** [comment]
:-1: The commit 779ca7799a71ebe84822b7e3dc02d922f14b2c54 (Job results: [913](https://jenkins.osquery.io/job/osqueryPullRequestFreeBSD/913/)) failed one or more tests (**FreeBSD**).

**Comment by osqueryer** [comment]
:-1: The commit 779ca7799a71ebe84822b7e3dc02d922f14b2c54 (Job results: [914](https://jenkins.osquery.io/job/osqueryPullRequestFreeBSD/914/)) failed one or more tests (**FreeBSD**).

**Pull request merged**